dataprofiler/profilers/profile_builder.py

```diff
@@ -1402,104 +1451,143 @@ def tqdm(l):
         # Newly introduced features (python3.8) improves the situation
         sample_ids = np.array(sample_ids)
 
-        # Create structured profile objects
-        new_cols = set()
-        for col in data.columns:
-            if col not in self._profile:
-                self._profile[col] = StructuredColProfiler(
+        # Initialize data column index -> _profile index mapping
+        col_idx_to_prof_idx = dict()
+
+        # Create StructuredColProfilers upon initialization
+        # Record correlation between columns in data and index in _profile
+        new_cols = False
+        if not self._initialized:
+            for col_idx in range(data.shape[1]):
+                col_name = data.columns[col_idx]
+                if isinstance(col_name, str):
+                    col_name = col_name.lower()
+                # Record index in _profile corresponding to current col profile
+                self._col_name_to_idx.setdefault(col_name, []).append(
+                    len(self._profile))
+                # Record where this column of data corresponds to _profile id
+                col_idx_to_prof_idx[col_idx] = len(self._profile)
+                # Add blank StructuredColProfiler to _profile
+                self._profile.append(StructuredColProfiler(
                     sample_size=sample_size,
                     min_true_samples=min_true_samples,
                     sample_ids=sample_ids,
                     options=self.options
-                )
-                new_cols.add(col)
-                
+                ))
+                new_cols = True
+        # Determine data column index -> profile index for unique col case
+        else:
+            # If duplicates present, schema enforced, so col_idx = prof_idx
+            col_idx_to_prof_idx = {i: i for i in range(len(self._profile))}
+            if not self._duplicate_cols_present:
+                # If columns unique, can allow permutation
+                for col_idx in range(data.shape[1]):
+                    col_name = data.columns[col_idx]
+                    if isinstance(col_name, str):
+                        col_name = col_name.lower()
+                    col_idx_to_prof_idx[col_idx] = \
+                        self._col_name_to_idx[col_name][0]
+
```

Comment:
I think if we utilize the mapping_given prior we may not need this at all.
e.g.
  1. Get mapping of new dataset (in case of add get mapping of other)
  2. Send to func to get dataset (other) to profile (self) mapping (raise error if schemas differ)
  3. If not initialized, create self._profile col profilers (not in add)
  4. Loop through using the mapping for updating etc.  (in add, just add based on mapping)